### PR TITLE
[Merged by Bors] - Refine and test slashing protection semantics

### LINKED ISSUE #1873: Fix bug in slashing protection import
## Description

There's an issue with our current implementation of slashing protection import in that it doesn't safely handle the case where 2+ interchange files are imported. E.g. if we import a file with history for epochs 0-50, and then another file with history for epochs 100-150, we still allow messages to be signed between 50-100, even though that violates the minimum bound rules for the 2nd file.

I've thought of two ways to fix this:

1. Add an explicit minimum bound column to the database, that gets populated only if we've imported an interchange file. I had experimented with this approach previously, but I didn't particularly like it because it adds specialised logic just for interchange files.
2. Prune the database upon import so that we discard any messages below the minimum bound from the latest file. This allows us to keep the validation logic the same (and consistent), but is potentially risky in that it involves deleting previously signed messages.

These two approaches are pretty much equivalent, but differ in whether the old messages actually get deleted from disk. I'm leaning towards (2) currently, so long as it is robustly tested.

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Closes #1873

## Proposed Changes

Fixes the bug in slashing protection import (#1873) by pruning the database upon import.

Also expands the test generator to cover this case and a few others which are under discussion here:

https://ethereum-magicians.org/t/eip-3076-validator-client-interchange-format-slashing-protection/4883

## Additional Info

Depending on the outcome of the discussion on Eth Magicians, we can either wait for consensus before merging, or merge our preferred solution and patch things later.

### CODE REVIEW

File: validator_client/slashing_protection/src/slashing_database.rs [hunk 1]
```diff
@@ -691,6 +714,48 @@ impl SlashingDatabase {
...
+        txn.execute(
+            "DELETE FROM signed_blocks
+             WHERE validator_id = ?1 AND slot < ?2",
+            params![validator_id, new_min_slot],
+        )?;
+
+        Ok(())
+    }
+
+    pub fn prune_signed_attestations(
```
Comment (michaelsproul) [inline validator_client/slashing_protection/src/slashing_database.rs:735]: Note to self: needs a comment

**Comment by michaelsproul** [comment]
This is now ready for review.

Guide for reviewers: the most important parts of this PR are the changes to `import_interchange_info`. The main changes are:

* Prune the signed_blocks and signed_attestations to the low watermark from the interchange file in order to solve the bug raised in #1873.
* Allow partial import of an interchange file: ignore records that appear slashable wrt to the existing database, or the interchange file itself. This still leaves some sharp corners on the UX for re-imports, but they will have to be addressed in subsequent work. I've taken the approach of making the code as safe as possible while changing as little as possible, i.e. not introducing drastic changes just before mainnet. Some other interchange implementers are pursuing an import strategy that allows seemingly slashable data to be imported (with better UX + less manual intervention), but I'm hesitant to put this into Lighthouse without first thoroughly testing all of the corner cases.
* Update the CLI command to output some summary information, so the user has some idea of the state of their database after the import, and direct them to a page in the Lighthouse book (and Discord) if they hit one of the sharp corners on import.

**Review by paulhauner — APPROVED** [review]
Looks great, I read thought it twice and didn't find anything. I did skip the tests though.

Merge at will :)

**Pull request closed without merging**